mitiq/executor/executor.py

```diff
@@ -160,38 +185,55 @@ def evaluate(
                 for circuit_with_measurements in observable.measure_in(circuit)
             ]
             result_step = observable.ngroups
-        elif (
-            observable is not None
-            and self._executor_return_type not in MeasurementResultLike
-            and self._executor_return_type not in DensityMatrixLike
-        ):
-            raise ValueError(
-                """Executor and observable are not compatible. Executors
-                returning expectation values as float must be used with
-                observable=None"""
-            )
         else:
             all_circuits = circuits
             result_step = 1
 
         # Run all required circuits.
-        all_results = self.run(all_circuits, force_run_all, **kwargs)
+        try:
+            all_results = self.run(all_circuits, force_run_all, **kwargs)
+        except Exception:
```

Comment:
Catch specific exceptions when possible. What fails in the above?

Reply:
We definitely should not be re-running on behalf of the user (imagine they're paying for hardware time). If something fails we need to let the user know, and let them fix the issue before continuing.

Reply:
There does not seem to be a way to assume the user has the executor set correctly without running the executor. As this could result in wasted runs and money, it seems like typehinting is necessary.

Reply:
As a general theme for the PR I think we should take the typehint as a source of truth if it is available, and fall back on whatever the use has supplied via function arguments otherwise and assume compatibility.

Reply:
I believe then that the main update for this PR will be explicitly checking incompatible cases only (float with observable, density matrix or measurement without observable.) At least for the density matrix and measurement this will throw an error BEFORE running the user defined executor and then trying to parse the results (when those would have failed.)

Reply:
IMO that's still a nice improvement, and can potentially save users some money/time.

Reply:
Agreed. I corrected this in https://github.com/unitaryfund/mitiq/pull/2514/commits/e3d8e2a7c3a9a8ab4516194e8429d54071656738